### PR TITLE
V2.0.0 kowhai login

### PR DESCRIPTION
**Pull request by red010b37**
Adds the login api - accepts the username and password and returns a jwt

### CODE REVIEW

**Comment by red010b37** [comment]
Also this pr is for the whole epic - oops 
https://github.com/NAVCoin/navpi-go/issues/66

So it includes the following tickets
https://github.com/NAVCoin/navpi-go/issues/59
https://github.com/NAVCoin/navpi-go/issues/60
https://github.com/NAVCoin/navpi-go/issues/61

File: app/daemon/daemon.go [hunk 1]
```diff
@@ -281,12 +286,17 @@ func getOSInfo() OSInfo {
...
 func downloadDaemon(serverConf conf.ServerConfig, version string) {
 
+
 	releaseInfo, _ := getReleaseDataForVersion(serverConf.ReleaseAPI, version)
 
 	dlPath, dlName, _ := getDownloadPathAndName(releaseInfo)
 
+	isGettingDaemon = true // flag we are getting the daemon
+
 	fs.DownloadExtract(dlPath, dlName)
```
Comment (rex-kwon) [inline app/daemon/daemon.go]: @red010b37 should we check for `err` after `fs.DownloadExtract`?

Comment (red010b37) [reply app/daemon/daemon.go]: @rex-kwon, no not at this stage as I think i want it to retry and not fatally error. I have found that the github api is rate limited. So need to look at caching that before we sort out the errors.

Comment (rex-kwon) [reply app/daemon/daemon.go]: sounds good

**Review by zanuka — APPROVED** [review]

File: app/user/userapi_test.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+			t.Error(err.Error())
+		}
+
+		assert.Equal(t, r.Code, http.StatusOK)
+		assert.NotEqual(t, len(apiResp.Data.(string)), 0)
+
+		//base test we have a jwt
+		split := strings.Split(apiResp.Data.(string), ".")
+		assert.NotEqual(t, len(split), 2)
+
```
Comment (rex-kwon) [inline app/user/userapi_test.go]: i like the gofight usage in tests 👍  how does the above assert test we have a jwt?

File: app/user/userapi_test.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+		// should be a 400
+		assert.Equal(t, r.Code, http.StatusBadRequest)
+
+		// should have 1 error
+		assert.Equal(t, len(apiResp.Errors), 1)
+
+		// check it has the right login error
+		resErr := apiResp.Errors[0]
+		assert.Equal(t, resErr.Code, api.AppRespErrors.LoginError.Code)
+
```
Comment (rex-kwon) [inline app/user/userapi_test.go]: nice tests 👊

File: app/user/userapi_test.go (hunk 2 above)
Comment (red010b37) [reply app/user/userapi_test.go]: JWT, should come in 3 parts separated by '.'

Comment (red010b37) [reply app/user/userapi_test.go]: Actually you were right, I have updated the test now it actually does a valid JWT test,

when i have write access I will commit

Comment (rex-kwon) [reply app/user/userapi_test.go]: ahh yeah that makes sense... cool cool, looking forward to jamming on this stuff tomorrow

**Pull request merged**